### ISSUE TITLE
Error: Could not load the default credentials shows up randomly once a day

### ISSUE THREAD

**Issue by gaofanmichael** [googleapis/nodejs-logging-winston#1]
_From @darciusal on August 2, 2017 18:58_

I have set up a google cloud function to monitor a bucket with frequent file changes.
The function reads files starting with a specific prefix and later generate new files to another bucket base on the files it reads, otherwise it will ignore the file and finish execution.
The function uses service account json to access/write files.
Those files starting with a specific prefix are updated only once a day.
The function is generating the new files properly and information seems correct.
However, by looking at the function logs, I notice that this error
 (Error: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information....) shows up once or twice a day. 
Could it be a bug in the gcloud library? 

_Copied from original issue: GoogleCloudPlatform/google-cloud-node#2501_

**Comment by gaofanmichael** [comment]
_From @darciusal on August 2, 2017 22:19_

I believe it's triggered by logging-winston which leads to the code here
https://github.com/google/google-auth-library-nodejs/blob/bd13376281a3af6e73fb729ea867950739c3c899/ts/lib/auth/googleauth.ts#208

**Comment by gaofanmichael** [comment]
_From @stephenplusplus on August 4, 2017 21:33_

@ofrobots any thoughts on this one?

**Comment by gaofanmichael** [comment]
_From @ofrobots on August 8, 2017 0:47_

Hi @darciusal! A few questions for you: What version of `google-cloud/logging-winston` are you using on your GCF application? Are there any other `google-cloud` libraries you are also using? Is there anything else functionally wrong in your function, or do you believe it is just the log messages that are being affected? The 'once or twice' errors you get, are they around a specific time, or at random times during the day.

I am trying to figure out if problem may be related to key expiry or whether it might be related to TCP timeouts in case your function got spun down to very low resources. Responses to the questions above might give me some clues.

**Comment by gaofanmichael** [comment]
_From @darciusal on August 11, 2017 6:25_

Hey @ofrobots,
I tried disabling logging-winston, but I was still getting this error.
I also encountered Auth error:Error: read ECONNRESET everyday.
For the function logic part, I have tried my best on making sure it's right. I have shared the function with you internally so you can confirm it as well.

Thanks!

**Comment by gaofanmichael** [comment]
_From @ofrobots on August 14, 2017 18:21_

Hi. I am pretty sure this issue is due to your Cloud Function getting throttled down to zero during periods of inactivity. Any TCP sockets that are pending at the time this happens will probably timeout/reset and error when the function is resumed on a subsequent request.

There are two different symptoms you reported here:
* `Error: Could not load the default credentials.`: This is reported by the `google-auth-library` (as you indicated above) which is used internally by both `storage` and `logging`. Subsequent requests to authenticate should succeed as the higher level libraries (e.g. logging) retry upon failure. Unless, of course, your function gets throttled again. I have opened https://github.com/google/google-auth-library-nodejs/issues/160 to get the `auth-library` to internally retry on socket timeout / connection reset.
* `Auth error:Error: read ECONNRESET`: This is coming from the gRPC library which is used by the logging library internally. It suggests the same throttling issue.

Fundamentally this means that, if your function is getting low enough traffic, the background work in your function may get starved due to throttling and as a result you may see those warnings.

One work-around you can try is by adding a delay before you terminate your function: 

```js
function eventHandler(event, callback) {
  // Your business logic here.
  // ...
  
  // delay terminating the function by a bit so that background tasks can drain.
 setTimeout(callback, 2000);
}
```

**Comment by gaofanmichael** [comment]
_From @darciusal on August 14, 2017 21:2_

Hi @ofrobots, Thanks for the response!
I have question on what do you mean for the "background work" you mentioned above.
Do you mean some of the business logic code that execute after the callback function is called?
If so, my function won't call the callback function until all logic are finish executing. (Example below)
```
const workPromise = createPromise();
workPromise.then(() => {...; callback()})
```

I have also noticed the ECONNRESET error before I used the winston function, so I am not sure if it is caused by the logging library.

Any chance do you think it is similar issue to this?
https://github.com/GoogleCloudPlatform/google-cloud-node/issues/2254
I thought they already changed the default forever agent to false though.

**Comment by stephenplusplus** [comment]
I'm going to close, since it appears the issue opener is no longer a registered GitHub member. If anyone else is still experiencing this issue, or I'm overlooking work that needs to be done, please let me know.

**Comment by opyate** [comment]
@stephenplusplus I'll assume the role of the issue opener :-)

I'm auditing all our triggers using `@google-cloud/logging-winston` and the log events will seemingly randomly not appear in StackDriver.

Please see this example where the logger's event listener sees the log messages which ought to go to StackDriver (they don't):

<img width="1567" alt="screen shot 2018-06-27 at 10 56 45" src="https://user-images.githubusercontent.com/96808/41968096-4d01d2a2-79fb-11e8-8ae3-44159e6e349b.png">

...and then a little bit later in StackDriver you'll see the OP's error:

<img width="1573" alt="screen shot 2018-06-27 at 10 57 52" src="https://user-images.githubusercontent.com/96808/41968102-5241bb4c-79fb-11e8-9bc8-589341763b6a.png">

I'll try and put @ofrobots suggestion of a "wait to drain" at the end of each trigger, but this is obviously not a long-term solution.

**Comment by opyate** [comment]
```
"winston": "^2.4.2"
"@google-cloud/logging-winston": "^0.9.0",
```

**Comment by ofrobots** [comment]
Hi @opyate. Thanks for the report and the detailed information. Based on the error, it seems that, for whatever reason, the code fails to identify the running environment to be GCF (or rather, GCP in general).  This causes auth to fail with the above error.

To be able to debug this further, I think we need to make a change in dependent module (gcp-metadata) that is currently swallowing unexpected errors instead of bubbling them up. I'm trying to fix this in https://github.com/stephenplusplus/gcp-metadata/pull/57, and once that gets released, it would good if you could try with the updated code. I want to know the precise reason why we were unable to identify your GCF running environment as lacking the metadata service.

I'll ping this thread once the modules have been updated with the necessary fix to debug this further.

**Comment by ofrobots** [comment]
@opyate  Looking at the OP, I think the error you're seeing is _slightly_ different. You running into an 'Unexpected error occurred...' which wraps the 'Unable to load application default' error. This is a slightly different path and I believe is a different root cause.

I am going to spin this off into a separate issue.

**Comment by opyate** [comment]
@ofrobots @stephenplusplus do you perhaps think I have missing log entries in StackDriver because I'm calling the logger in a `continuation-local-storage` context?

I.e. this portion of code:

```
return session.runAndReturn(() => {
            session.set('correlation-id', uuidv4())
            logger.info(`[audit pre] ${action}`, Object.assign(payload, {result: 'running...'}))

            return cb(...args).then(() => {
                return logger.info(`[audit] ${action}`, Object.assign(payload, {result: 'success'}))
            }).catch(err => {
                logger.error(`[audit] ${action}`, Object.assign(payload, {result: 'error'}))
                throw err
            })
        })
```
